File: src/powerpointGenerator.js
/**
 * PowerPoint Generator
 * Creates PowerPoint presentations with sprint summaries in 2x2 format
 */

import pptxgen from 'pptxgenjs';
import fs from 'fs';
import path from 'path';
import { LLMSummaryGenerator } from './llmSummaryGenerator.js';

export class PowerPointGenerator {
  constructor(provider, apiKey, model) {
    this.pptx = new pptxgen();
    this.pptx.author = 'Sprint Summary Agent';
    this.pptx.company = 'International SOS';
    this.pptx.title = 'Sprint Summary Report';
    this.llmGenerator = new LLMSummaryGenerator(provider, apiKey, model);
  }

  /**
   * Create a 2x2 slide for a team's sprint summary with LLM-generated content
   */
  async createTeamSlide(summary, sprintData, metrics) {
    const slide = this.pptx.addSlide();
    const { sprintInfo, projectInfo, teamInfo, sprintHealthMetrics,
            sprintHealthAnalysis, currentBlockers, keyAccomplishments } = summary;

    // Construct proper data structures for LLM
    const teamInfoForLLM = teamInfo || { label: sprintData.teamLabel };
    const sprintInfoForLLM = sprintInfo || sprintData.sprint;
    const projectInfoForLLM = projectInfo || sprintData.project;

    // Generate LLM content for the slide
    console.log(`   🤖 Generating AI slide content for ${teamInfoForLLM.label}...`);
    const slideContent = await this.llmGenerator.generateSlideContent(
      {
        sprintInfo: sprintInfoForLLM,
        projectInfo: projectInfoForLLM,
        teamInfo: teamInfoForLLM
      },
      metrics,
      sprintHealthAnalysis,
      currentBlockers,
      keyAccomplishments
    );

    // Title: Team Name and Sprint
    slide.addText(`${teamInfoForLLM.label.toUpperCase()} - ${sprintInfoForLLM.name}`, {
      x: 0.5,
      y: 0.3,
      w: 9,
      h: 0.5,
      fontSize: 28,
      bold: true,
      color: '363636',
      align: 'center'
    });

    // Subtitle: Project and Date Range
    const startDate = new Date(sprintInfoForLLM.startDate).toLocaleDateString();
    const endDate = new Date(sprintInfoForLLM.endDate).toLocaleDateString();
    slide.addText(`${projectInfoForLLM.name} | ${startDate} - ${endDate}`, {
      x: 0.5,
      y: 0.9,
      w: 9,
      h: 0.3,
      fontSize: 14,
      color: '666666',
      align: 'center'
    });

    // Define 2x2 grid layout - adjusted for better fit
    const boxWidth = 4.4;
    const boxHeight = 1.6;
    const leftX = 0.5;
    const rightX = 5.1;
    const topY = 1.4;
    const bottomY = 3.0;

    // Health color based on status
    const healthColors = {
      'Good': '4CAF50',
      'Fair': 'FFA726',
      'Poor': 'EF5350'
    };
    const healthColor = healthColors[sprintHealthMetrics.overallHealth] || '757575';

    // TOP LEFT: Sprint Health Metrics
    slide.addShape('rect', {
      x: leftX,
      y: topY,
      w: boxWidth,
      h: boxHeight,
      fill: { color: 'F5F5F5' },
      line: { color: 'DDDDDD', width: 1 }
    });

    slide.addText(`📊 ${slideContent.healthSummary.title}`, {
      x: leftX + 0.1,
      y: topY + 0.1,
      w: boxWidth - 0.2,
      h: 0.3,
      fontSize: 12,
      bold: true,
      color: '363636'
    });

    const healthBullets = slideContent.healthSummary.bullets.slice(0, 5).map(b => `• ${b}`).join('\n');

    slide.addText(healthBullets, {
      x: leftX + 0.1,
      y: topY + 0.45,
      w: boxWidth - 0.2,
      h: boxHeight - 0.55,
      fontSize: 9,
      color: '424242',
      valign: 'top'
    });

    // Health indicator circle
    slide.addShape('ellipse', {
      x: leftX + boxWidth - 0.8,
      y: topY + 0.15,
      w: 0.5,
      h: 0.5,
      fill: { color: healthColor }
    });

    // TOP RIGHT: Key Accomplishments
    slide.addShape('rect', {
      x: rightX,
      y: topY,
      w: boxWidth,
      h: boxHeight,
      fill: { color: 'F5F5F5' },
      line: { color: 'DDDDDD', width: 1 }
    });

    slide.addText(`🎯 ${slideContent.accomplishments.title}`, {
      x: rightX + 0.1,
      y: topY + 0.1,
      w: boxWidth - 0.2,
      h: 0.3,
      fontSize: 12,
      bold: true,
      color: '363636'
    });

    const accomplishmentsBullets = slideContent.accomplishments.bullets.slice(0, 5).map(b => `• ${b}`).join('\n');

    slide.addText(accomplishmentsBullets || 'No key accomplishments recorded', {
      x: rightX + 0.1,
      y: topY + 0.45,
      w: boxWidth - 0.2,
      h: boxHeight - 0.55,
      fontSize: 8,
      color: '424242',
      valign: 'top'
    });

    // BOTTOM LEFT: Current Blockers
    slide.addShape('rect', {
      x: leftX,
      y: bottomY,
      w: boxWidth,
      h: boxHeight,
      fill: { color: 'F5F5F5' },
      line: { color: 'DDDDDD', width: 1 }
    });

    slide.addText(`🚧 ${slideContent.blockers.title}`, {
      x: leftX + 0.1,
      y: bottomY + 0.1,
      w: boxWidth - 0.2,
      h: 0.3,
      fontSize: 12,
      bold: true,
      color: '363636'
    });

    const blockersBullets = slideContent.blockers.bullets.slice(0, 5).map(b => `• ${b}`).join('\n');
    const hasBlockers = currentBlockers.length > 0;

    slide.addText(blockersBullets, {
      x: leftX + 0.1,
      y: bottomY + 0.45,
      w: boxWidth - 0.2,
      h: boxHeight - 0.55,
      fontSize: 8,
      color: hasBlockers ? 'D32F2F' : '388E3C',
      valign: 'top'
    });

    // BOTTOM RIGHT: Recommendations
    slide.addShape('rect', {
      x: rightX,
      y: bottomY,
      w: boxWidth,
      h: boxHeight,
      fill: { color: 'F5F5F5' },
      line: { color: 'DDDDDD', width: 1 }
    });

    slide.addText(`💡 ${slideContent.recommendations.title}`, {
      x: rightX + 0.1,
      y: bottomY + 0.1,
      w: boxWidth - 0.2,
      h: 0.3,
      fontSize: 12,
      bold: true,
      color: '363636'
    });

    const recommendationsBullets = slideContent.recommendations.bullets.slice(0, 5).map(b => `• ${b}`).join('\n');

    slide.addText(recommendationsBullets, {
      x: rightX + 0.1,
      y: bottomY + 0.45,
      w: boxWidth - 0.2,
      h: boxHeight - 0.55,
      fontSize: 8,
      color: '424242',
      valign: 'top'
    });
  }

  /**
   * Create a summary slide with all teams
   */
  createOverviewSlide(allSummaries) {
    const slide = this.pptx.addSlide();

    // Title
    slide.addText('Sprint Summary Overview - All Teams', {
      x: 0.5,
      y: 0.3,
      w: 9,
      h: 0.5,
      fontSize: 32,
      bold: true,
      color: '363636',
      align: 'center'
    });

    // Subtitle
    const date = new Date().toLocaleDateString();
    slide.addText(`Generated: ${date}`, {
      x: 0.5,
      y: 0.9,
      w: 9,
      h: 0.3,
      fontSize: 14,
      color: '666666',
      align: 'center'
    });

    // Calculate aggregate metrics
    let totalIssues = 0;
    let completedIssues = 0;
    let totalPoints = 0;
    let completedPoints = 0;
    let totalBlockers = 0;

    allSummaries.forEach(summary => {
      totalIssues += summary.sprintHealthMetrics.totalIssues;
      completedIssues += summary.sprintHealthMetrics.completedIssues;
      totalPoints += summary.sprintHealthMetrics.totalStoryPoints;
      completedPoints += summary.sprintHealthMetrics.completedStoryPoints;
      totalBlockers += summary.sprintHealthMetrics.blockedIssues;
    });

    const overallCompletion = totalIssues > 0 ? ((completedIssues / totalIssues) * 100).toFixed(1) : 0;
    const overallVelocity = totalPoints > 0 ? ((completedPoints / totalPoints) * 100).toFixed(1) : 0;

    // Aggregate metrics box
    slide.addShape('rect', {
      x: 1,
      y: 1.5,
      w: 8,
      h: 2,
      fill: { color: 'E3F2FD' },
      line: { color: '1976D2', width: 2 }
    });

    const aggregateText = [
      `Total Teams: ${allSummaries.length}`,
      `Total Issues: ${completedIssues}/${totalIssues} (${overallCompletion}%)`,
      `Total Story Points: ${completedPoints}/${totalPoints} (${overallVelocity}%)`,
      `Total Blockers: ${totalBlockers}`
    ];

    slide.addText(aggregateText.join('\n'), {
      x: 1.5,
      y: 1.8,
      w: 7,
      h: 1.5,
      fontSize: 18,
      bold: true,
      color: '1565C0',
      valign: 'middle'
    });

    // Team summary table
    const tableData = [
      [
        { text: 'Team', options: { bold: true, color: 'FFFFFF', fill: '1976D2' } },
        { text: 'Sprint', options: { bold: true, color: 'FFFFFF', fill: '1976D2' } },
        { text: 'Health', options: { bold: true, color: 'FFFFFF', fill: '1976D2' } },
        { text: 'Completion', options: { bold: true, color: 'FFFFFF', fill: '1976D2' } },
        { text: 'Velocity', options: { bold: true, color: 'FFFFFF', fill: '1976D2' } }
      ]
    ];

    allSummaries.forEach(summary => {
      const healthColors = {
        'Good': '4CAF50',
        'Fair': 'FFA726',
        'Poor': 'EF5350'
      };
      const healthColor = healthColors[summary.sprintHealthMetrics.overallHealth] || '757575';

      tableData.push([
        summary.teamInfo.label.toUpperCase(),
        summary.sprintInfo.name,
        { text: summary.sprintHealthMetrics.overallHealth, options: { fill: healthColor, color: 'FFFFFF', bold: true } },
        summary.sprintHealthMetrics.completionRate,
        String(summary.sprintHealthMetrics.velocity)
      ]);
    });

    slide.addTable(tableData, {
      x: 1,
      y: 4,
      w: 8,
      h: 4,
      fontSize: 12,
      border: { pt: 1, color: 'DDDDDD' }
    });
  }

  /**
   * Generate PowerPoint presentation
   */
  async generatePresentation(allSummaries, allSprintData, allMetrics, outputDir = './output') {
    if (!fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir, { recursive: true });
    }

    // Create overview slide first
    this.createOverviewSlide(allSummaries);

    // Create a slide for each team (must process sequentially for LLM calls)
    for (let i = 0; i < allSummaries.length; i++) {
      await this.createTeamSlide(allSummaries[i], allSprintData[i], allMetrics[i]);
    }

    // Save presentation
    const filePath = path.join(outputDir, 'sprint-summary-presentation.pptx');
    await this.pptx.writeFile({ fileName: filePath });

    console.log(`PowerPoint presentation saved to: ${filePath}`);
    return filePath;
  }
}

File: sprint_summary_agent/powerpoint_generator.py
"""PowerPoint Generator - Simple version for maximum compatibility."""

from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List

from pptx import Presentation
from pptx.dml.color import RGBColor
from pptx.enum.text import PP_ALIGN, MSO_ANCHOR
from pptx.util import Inches, Pt

from .llm_summary_generator import LLMSummaryGenerator


class PowerPointGenerator:
    """Creates PowerPoint presentations with sprint summaries - simplified for compatibility."""

    # Color scheme
    COLORS = {
        "good": RGBColor(76, 175, 80),      # Green
        "fair": RGBColor(255, 167, 38),     # Orange
        "poor": RGBColor(239, 83, 80),      # Red
        "gray": RGBColor(117, 117, 117),    # Gray
        "dark_gray": RGBColor(54, 54, 54),  # Dark gray
        "light_gray": RGBColor(245, 245, 245),  # Light gray background
        "border": RGBColor(221, 221, 221),  # Border gray
        "text": RGBColor(66, 66, 66),       # Text gray
        "blocker_red": RGBColor(211, 47, 47),  # Blocker red
        "success_green": RGBColor(56, 142, 60),  # Success green
        "blue": RGBColor(33, 150, 243),     # Blue accent
    }

    def __init__(self, provider: str, api_key: str, model: str):
        """Initialize with LLM provider configuration."""
        self.prs = Presentation()
        self.prs.slide_width = Inches(10)
        self.prs.slide_height = Inches(7.5)
        self.llm_generator = LLMSummaryGenerator(provider, api_key, model)

    def generate_presentation(
        self,
        all_summaries: List[Dict[str, Any]],
        all_sprint_data: List[Dict[str, Any]],
        all_metrics: List[Dict[str, Any]],
        output_dir: str = "./output",
    ):
        """Generate PowerPoint presentation with LLM-powered content."""
        # Add title slide
        self._create_title_slide(all_summaries)

        # Add slide for each team
        for i, (summary, sprint_data, metrics) in enumerate(zip(all_summaries, all_sprint_data, all_metrics)):
            print(f"   🤖 Generating AI slide content for {summary['teamInfo']['label']}...")
            self._create_team_slide(summary, sprint_data, metrics)

        # Save presentation
        output_path = Path(output_dir)
        output_path.mkdir(parents=True, exist_ok=True)
        file_path = output_path / "sprint-summary-presentation.pptx"

        self.prs.save(str(file_path))
        print(f"   ✅ PowerPoint saved to: {file_path}")

    def _create_title_slide(self, all_summaries: List[Dict[str, Any]]):
        """Create title slide - simplified version."""
        title_layout = self.prs.slide_layouts[0]  # Use title slide layout
        slide = self.prs.slides.add_slide(title_layout)

        # Set title
        title = slide.shapes.title
        title.text = "📊 Sprint Summary Report"
        title.text_frame.paragraphs[0].font.size = Pt(44)
        title.text_frame.paragraphs[0].font.color.rgb = self.COLORS["blue"]

        # Set subtitle with projects and teams
        projects = sorted(set(s["projectInfo"]["key"] for s in all_summaries))
        teams = sorted(set(s["teamInfo"]["label"] for s in all_summaries))

        subtitle = slide.placeholders[1]
        subtitle.text = f"📁 Projects: {', '.join(projects)}\n👥 Teams: {', '.join(teams)}\n\nGenerated: {datetime.now().strftime('%Y-%m-%d %H:%M')}"
        for paragraph in subtitle.text_frame.paragraphs:
            paragraph.font.size = Pt(18)
            paragraph.font.color.rgb = self.COLORS["dark_gray"]

    def _create_team_slide(
        self,
        summary: Dict[str, Any],
        sprint_data: Dict[str, Any],
        metrics: Dict[str, Any],
    ):
        """Create a slide for a specific team with 2x2 layout - simplified."""
        blank_layout = self.prs.slide_layouts[6]
        slide = self.prs.slides.add_slide(blank_layout)

        team_label = summary["teamInfo"]["label"]
        project_key = summary["projectInfo"]["key"]
        health = summary["sprintHealthMetrics"]["overallHealth"]

        # Title at top
        title_box = slide.shapes.add_textbox(Inches(0.5), Inches(0.2), Inches(9), Inches(0.6))
        title_frame = title_box.text_frame
        title_frame.text = f"Team: {team_label} ({project_key}) - Health: {health}"
        title_para = title_frame.paragraphs[0]
        title_para.font.size = Pt(28)
        title_para.font.bold = True
        title_para.font.color.rgb = self._get_health_color(health)

        # Get LLM-generated content
        llm_content = self.llm_generator.generate_slide_content(summary)

        # 2x2 layout positions - use simple textboxes only
        positions = [
            (Inches(0.5), Inches(1.2)),  # Top left
            (Inches(5.2), Inches(1.2)),  # Top right
            (Inches(0.5), Inches(4.2)),  # Bottom left
            (Inches(5.2), Inches(4.2)),  # Bottom right
        ]

        box_width = Inches(4.3)
        box_height = Inches(2.8)

        # Top Left: Health Summary
        self._add_simple_box(
            slide,
            positions[0][0],
            positions[0][1],
            box_width,
            box_height,
            llm_content["healthSummary"]["title"],
            llm_content["healthSummary"]["bullets"],
            self._get_health_color(health),
        )

        # Top Right: Accomplishments
        self._add_simple_box(
            slide,
            positions[1][0],
            positions[1][1],
            box_width,
            box_height,
            llm_content["accomplishments"]["title"],
            llm_content["accomplishments"]["bullets"],
            self.COLORS["success_green"],
        )

        # Bottom Left: Blockers
        blocker_color = (
            self.COLORS["blocker_red"]
            if summary.get("currentBlockers") and len(summary["currentBlockers"]) > 0
            else self.COLORS["success_green"]
        )
        self._add_simple_box(
            slide,
            positions[2][0],
            positions[2][1],
            box_width,
            box_height,
            llm_content["blockers"]["title"],
            llm_content["blockers"]["bullets"],
            blocker_color,
        )

        # Bottom Right: Recommendations
        self._add_simple_box(
            slide,
            positions[3][0],
            positions[3][1],
            box_width,
            box_height,
            llm_content["recommendations"]["title"],
            llm_content["recommendations"]["bullets"],
            self.COLORS["blue"],
        )

    def _add_simple_box(
        self,
        slide,
        left,
        top,
        width,
        height,
        title,
        bullets,
        title_color,
    ):
        """Add a simple text box - no shapes, just text for compatibility."""
        # Create textbox for entire area
        textbox = slide.shapes.add_textbox(left, top, width, height)
        text_frame = textbox.text_frame
        text_frame.word_wrap = True
        text_frame.margin_top = Pt(10)
        text_frame.margin_left = Pt(10)
        text_frame.margin_right = Pt(10)

        # Add title
        p = text_frame.paragraphs[0]
        p.text = title
        p.font.size = Pt(16)
        p.font.bold = True
        p.font.color.rgb = title_color
        p.space_after = Pt(8)

        # Add bullets
        for bullet in bullets:
            p = text_frame.add_paragraph()
            p.text = f"• {bullet}"
            p.font.size = Pt(11)
            p.font.color.rgb = self.COLORS["text"]
            p.space_before = Pt(4)

    def _get_health_color(self, health: str) -> RGBColor:
        """Get color for health status."""
        health_lower = health.lower()
        if health_lower == "good":
            return self.COLORS["good"]
        elif health_lower == "fair":
            return self.COLORS["fair"]
        elif health_lower == "poor":
            return self.COLORS["poor"]
        else:
            return self.COLORS["gray"]
